### PR TITLE
feat(ui): scope repository selector by peer

### PR DESCRIPTION
**Pull request by rudolfs**
feat

* Move source browsing to a querystring implementation
* Add mock default_peer attribute to Project
* Add peer scope to source browsing
* Remove revisions title
* Move all source browsing state into the Source screen
* Align peer and branch in revision selector
* Fetch tree data just in time when it's needed
* Make File component leaner
* Select mock peer in revision selector
* Add spec for repo browser when switching between projects

Closes: #443.

### CODE REVIEW

File: ui/DesignSystem/Component/SourceBrowser/CommitTeaser.svelte [hunk 1]
```diff
@@ -59,7 +60,7 @@
     <Icon.Commit style="fill: var(--color-secondary)" />
     <a
       class="commit-sha"
-      href={path.projectCommit(projectId, commitSha)}
+      href={path.projectCommit(projectId, userId, commitSha)}
```
Comment (xla) [inline ui/DesignSystem/Component/SourceBrowser/CommitTeaser.svelte:63]: Why does this depend on user information?

File: ui/src/source.ts [hunk 2]
```diff
@@ -118,18 +112,20 @@ enum Kind {
...
   FetchCommits = "FETCH_COMMITS",
   FetchRevisions = "FETCH_REVISIONS",
-  Update = "UPDATE",
+  FetchObject = "FETCH_OBJECT",
 }
 
 interface FetchCommit extends event.Event<Kind> {
   kind: Kind.FetchCommit;
   projectId: string;
+  userId: string;
```
Comment (xla) [inline ui/src/source.ts:121]:
```suggestion
  userUrn: string;
```

File: ui/src/source.ts [hunk 3]
```diff
@@ -118,18 +112,20 @@ enum Kind {
...
   sha1: string;
 }
 
 interface FetchCommits extends event.Event<Kind> {
   kind: Kind.FetchCommits;
   projectId: string;
+  userId: string;
```
Comment (xla) [inline ui/src/source.ts:128]:
```suggestion
  userUrn: string;
```

File: ui/src/source.ts [hunk 4]
```diff
@@ -138,9 +134,10 @@ interface FetchRevisions extends event.Event<Kind> {
   projectId: string;
 }
 
-interface Update extends event.Event<Kind> {
-  kind: Kind.Update;
+interface FetchObject extends event.Event<Kind> {
+  kind: Kind.FetchObject;
   path: string;
+  user: string;
```
Comment (xla) [inline ui/src/source.ts:140]:
```suggestion
  userUrn: string;
```

File: ui/DesignSystem/Component/SourceBrowser/CommitTeaser.svelte (hunk 1 above)
Comment (cloudhead) [reply ui/DesignSystem/Component/SourceBrowser/CommitTeaser.svelte:63]: I think it's the `PeerId`?

File: ui/src/source.ts (hunk 3 above)
Comment (cloudhead) [reply ui/src/source.ts:128]: I don't think that's correct, it's the `PeerId` if I'm not mistaken. But not sure that's what we want to call it in the UI components..

File: proxy/src/project.rs [hunk 5]
```diff
@@ -50,21 +52,26 @@ pub struct Project {
     pub registration: Option<Registration>,
     /// High-level statistics about the project
     pub stats: coco::Stats,
+    /// The user to which all source browsing queries are scoped to.
+    pub default_user: identity::Identity,
```
Comment (cloudhead) [inline proxy/src/project.rs:56]: Does this belong here? I would think it's part of the Metadata (since that's where we get it from)?

Comment (xla) [reply proxy/src/project.rs:56]: The metadata of a project is the bucket for user defined input during project creation and update. It's not derived or generated data.

File: ui/Screen/Project/Source.svelte [hunk 6]
```diff
@@ -221,8 +238,8 @@
         {#if object.info.objectType === ObjectType.Blob}
           <FileSource
             blob={object}
-            path={$currentPath}
-            rootPath={path.projectSource(project.id)}
+            path={currentObjectPath}
+            rootPath={path.projectSource(project.id, project.default_user.id)}
```
Comment (cloudhead) [inline ui/Screen/Project/Source.svelte:242]: So I'm guessing this PR doesn't scope files to the currently selected peer? Just commits?

File: ui/src/path.ts [hunk 7]
```diff
@@ -37,24 +32,41 @@ export const projectIssues = (id: string): string => `/projects/${id}/issues`;
...
-export const projectCommit = (id: string, hash: string): string =>
-  `/projects/${id}/commit/${hash}`;
-export const projectCommits = (id: string, branch: string): string =>
-  `/projects/${id}/commits/${encodeURIComponent(branch)}`;
+
+export const projectCommit = (id: string, user: string, hash: string): string =>
+  `/projects/${id}/${user}/commit/${hash}`;
+export const projectCommits = (
+  id: string,
+  user: string,
```
Comment (cloudhead) [inline ui/src/path.ts:67]: Let's figure out what this "user" is: If it's a URN, eg. `rad:git:12389asd8`, then `user` is appropriate. If it's a PeerID on the other hand, we should call this `peer`.

File: ui/Screen/Org/Projects.svelte [hunk 8]
```diff
@@ -22,7 +22,12 @@
   const select = event => {
     const orgProject = event.detail;
     if (orgProject.maybeProject) {
-      push(path.projectSource(orgProject.maybeProject.id));
+      push(
+        path.projectSource(
+          orgProject.maybeProject.id,
+          orgProject.maybeProject.default_user.id
```
Comment (cloudhead) [inline ui/Screen/Org/Projects.svelte:28]: So here it looks like the project source indeed depends on a user. But I was under the impression that source code is tied to the device/peer, not the user. How do we go from `user -> device -> repo`?

File: proxy/src/project.rs (hunk 5 above)
Comment (cloudhead) [reply proxy/src/project.rs:56]: Ah I see, so it's not the project identity.

File: ui/Screen/Org/Projects.svelte (hunk 8 above)
Comment (xla) [reply ui/Screen/Org/Projects.svelte:28]: Same concern as in https://github.com/radicle-dev/radicle-upstream/pull/620#pullrequestreview-440583382

File: ui/DesignSystem/Component/SourceBrowser/CommitTeaser.svelte (hunk 1 above)
Comment (rudolfs) [reply ui/DesignSystem/Component/SourceBrowser/CommitTeaser.svelte:63]: Yeah, it'll be the `PeerId` for now, but I thought that we'd be able to resolve them to user's soon, is that not correct?

**Comment by rudolfs** [comment]
> Looks like we're almost there, but like @xla I'm a little confused about the user being so closely tied into this, given that if I'm not mistaken, source code is tied to peers.

My thinking on this was that since we'd want to resolve `PeerId`s to users in the UI, I'd only handle that in the UI and expect the resolution to happen in the backend. But maybe that's premature since the whole peer <-> user resolution is still being worked out.

Happy to switch out everything for a `PeerId`. @cloudhead should we have a short chat about this first?

File: ui/Screen/Project/Source.svelte (hunk 6 above)
Comment (rudolfs) [reply ui/Screen/Project/Source.svelte:242]: Indeed, changing the remaining endpoints to accept a `userId` (or `peerId`) is still TBD. Wanted to get the UI routing merged first before @xla rewrites the whole router...

**Comment by xla** [comment]
> My thinking on this was that since we'd want to resolve `PeerId`s to users in the UI, I'd only handle that in the UI and expect the resolution to happen in the backend. But maybe that's premature since the whole peer <-> user resolution is still being worked out.

While the resolution indeed should happen transparently, there still is the reality that there is a 1:N to relationship between `Project` and `Repository`/source tree, where the later is tight to a `PeerID`. In turn that means that we can end up with many repositories from the same user, but from different peers. @FintanH Correct me if I'm wrong here.

> Happy to switch out everything for a `PeerId`. @cloudhead should we have a short chat about this first?

For now let's simplify and just use peerids in the first iteration and make sure to inform the data model in the proxy from the hiereachy I outline above.

File: ui/src/path.ts (hunk 7 above)
Comment (FintanH) [reply ui/src/path.ts:67]: I think `userUrn` would be more helpful. `user: String` doesn't portray much to me, it could just be the handle for all I know (*cough* types *cough*). So I think `userUrn` or `peerId` would more indicative.

**Comment by FintanH** [comment]
> While the resolution indeed should happen transparently, there still is the reality that there is a 1:N to relationship between Project and Repository/source tree, where the later is tight to a PeerID. In turn that means that we can end up with many repositories from the same user, but from different peers. @FintanH Correct me if I'm wrong here.

Yup, this is in similar vein to my comment here https://github.com/radicle-dev/radicle-upstream/pull/620#discussion_r448214355. We have 1:N for User devices (technically you could have multiple User's per device but let's not get into _that_ :)). This means that you may want to look at `fintohaps@yfijf8575` _or_ `fintohap@iuhd8485` because they use two machines and they both show up in the selector.

File: ui/Screen/Project/Source.svelte (hunk 6 above)
Comment (rudolfs) [reply ui/Screen/Project/Source.svelte]: The remaining endpoints are taken care of in:
https://github.com/radicle-dev/radicle-upstream/pull/620/commits/e2c4a7f0d339a145ca4c94b8b18cd57116b897f8
https://github.com/radicle-dev/radicle-upstream/pull/620/commits/d4f044e705b391b371409455a496b9941fedb075
https://github.com/radicle-dev/radicle-upstream/pull/620/commits/f25eb586bf32e7ddc3abd70fce13cdd2466af24c
https://github.com/radicle-dev/radicle-upstream/pull/620/commits/f25eb586bf32e7ddc3abd70fce13cdd2466af24c

File: ui/DesignSystem/Component/SourceBrowser/CommitTeaser.svelte (hunk 1 above)
Comment (rudolfs) [reply ui/DesignSystem/Component/SourceBrowser/CommitTeaser.svelte]: @cloudhead addressed in https://github.com/radicle-dev/radicle-upstream/pull/620/commits/d5b10c171dddbf5b93c4a3b1493d25c1689799f0

@FintanH no, this just constructs a link to the single commit screen, when you click the link the screen will fetch the respective commit and use these params as scope.

Screenshot 2020-07-01 at 16 21 40

File: ui/src/source.ts (hunk 2 above)
Comment (rudolfs) [reply ui/src/source.ts]: https://github.com/radicle-dev/radicle-upstream/pull/620/commits/d5b10c171dddbf5b93c4a3b1493d25c1689799f0

File: ui/src/source.ts (hunk 3 above)
Comment (rudolfs) [reply ui/src/source.ts]: https://github.com/radicle-dev/radicle-upstream/pull/620/commits/d5b10c171dddbf5b93c4a3b1493d25c1689799f0

File: ui/src/source.ts (hunk 4 above)
Comment (rudolfs) [reply ui/src/source.ts]: https://github.com/radicle-dev/radicle-upstream/pull/620/commits/d5b10c171dddbf5b93c4a3b1493d25c1689799f0

File: proxy/src/project.rs (hunk 5 above)
Comment (rudolfs) [reply proxy/src/project.rs]: Had a chat with @xla IRL, we decided that the ordering of the `GET /source/revisions` response (which also contains peers) will determine the default_peer. It's going to be the first in the list.

File: ui/src/path.ts (hunk 7 above)
Comment (rudolfs) [reply ui/src/path.ts]: https://github.com/radicle-dev/radicle-upstream/pull/620/commits/d5b10c171dddbf5b93c4a3b1493d25c1689799f0

File: ui/Screen/Org/Projects.svelte (hunk 8 above)
Comment (rudolfs) [reply ui/Screen/Org/Projects.svelte]: https://github.com/radicle-dev/radicle-upstream/pull/620/commits/d5b10c171dddbf5b93c4a3b1493d25c1689799f0

File: ui/DesignSystem/Component/SourceBrowser/CommitTeaser.svelte (hunk 1 above)
Comment (FintanH) [reply ui/DesignSystem/Component/SourceBrowser/CommitTeaser.svelte]: > I think it is fine to be able to jump to any commit, as long as some metadata is computed which shows in which projects + branches it appears. The only ambiguous case is when you jump to a commit in the current project, but that commit doesn't appear in that project anywhere.

As Kim said, it might not be that we scope it. I would suggest we remove this change until we're sure what the behaviour will be.

Comment (rudolfs) [reply ui/DesignSystem/Component/SourceBrowser/CommitTeaser.svelte]: Alright.

Comment (rudolfs) [reply ui/DesignSystem/Component/SourceBrowser/CommitTeaser.svelte]: @FintanH https://github.com/radicle-dev/radicle-upstream/pull/620/commits/cc7bc8a3819ab9ab419487c80f0758ca56898462

File: proxy/src/http/source.rs [hunk 9]
```diff
@@ -274,16 +276,28 @@ mod handler {
 
     /// Fetch a [`coco::Blob`].
     pub async fn blob(
-        peer: Arc<Mutex<coco::PeerApi>>,
+        api: Arc<Mutex<coco::PeerApi>>,
```
Comment (FintanH) [inline proxy/src/http/source.rs:279]: Was this @xla changing the naming? xD

File: proxy/src/coco/source.rs [hunk 10]
```diff
@@ -197,6 +198,7 @@ pub struct TreeEntry {
 /// Will return [`error::Error`] if the project doesn't exist or a surf interaction fails.
 pub fn blob(
     browser: &mut Browser,
+    _peer_id: Option<&peer::PeerId>,
```
Comment (FintanH) [inline proxy/src/coco/source.rs:201]: I'm unsure if we need the `PeerId` at this scope, but that can be determined in another PR.

File: proxy/src/http/source.rs [hunk 11]
```diff
@@ -795,6 +835,7 @@ mod test {
         let want = coco::with_browser(&*peer.lock().await, &urn, |mut browser| {
             coco::blob(
                 &mut browser,
+                None,
```
Comment (FintanH) [inline proxy/src/http/source.rs:838]: Could use `peer.peed_id` or the peer id derived from the key. Maybe for a later change.

File: ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte [hunk 12]
```diff
@@ -108,10 +114,11 @@
   hidden={expanded}>
   <div class="selector-avatar">
     <Avatar
-      title={revisions[0].identity.metadata.handle}
-      avatarFallback={revisions[0].identity.avatarFallback}
+      title={currentSelectedUser.identity.metadata.handle}
```
[the remark this answers is not in the document]
Comment (FintanH) [reply ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte]: This says it's outdated but I don't see the change I was expecting :eyes:

File: proxy/src/http/source.rs (hunk 9 above)
Comment (rudolfs) [reply proxy/src/http/source.rs:279]: Indeed, he was : D

File: ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte (hunk 12 above)
[the remark this answers is not in the document]
Comment (rudolfs) [reply ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte]: It says outdated, because the underlying code changed in the user -> peer rename commit. Haven't yet had time to address this comment.

**Comment by rudolfs** [comment]
> The concept of a "default user" for a project seems odd. What is a non-default user? What is a user in general in the context of a project? Isn't that we have multiple source trees for a project, namespaced by peerids which can resolve to user urns?

@xla https://github.com/radicle-dev/radicle-upstream/pull/620/commits/06e9bdacf9c31c800b28cf6779cb1559d4323175

File: proxy/src/project.rs (hunk 5 above)
Comment (rudolfs) [reply proxy/src/project.rs]: https://github.com/radicle-dev/radicle-upstream/pull/620/commits/06e9bdacf9c31c800b28cf6779cb1559d4323175

File: proxy/src/project.rs [hunk 13]
```diff
@@ -81,10 +88,29 @@ pub enum Registration {
...
+    }
+}
+
 /// Fetch the project with a given urn from a peer
 pub fn get(peer: &coco::PeerApi, project_urn: &uri::RadUrn) -> Result<Project, error::Error> {
     let project = coco::get_project(peer, project_urn)?;
     let stats = coco::with_browser(peer, project_urn, |browser| Ok(browser.get_stats()?))?;
 
-    Ok((project, stats).into())
+    Ok((project, stats, get_fake_default_user()).into())
```
[the remark this answers is not in the document]
Comment (rudolfs) [reply proxy/src/project.rs]: The default_user got removed. I think we should tackle the proxy end of this in https://github.com/radicle-dev/radicle-upstream/pull/510.

File: cypress/integration/project_source_browsing.spec.js [hunk 14]
```diff
@@ -196,6 +196,17 @@ context("source code browsing", () => {
         cy.contains("test-file-deletion.txt").should("not.exist");
       });
     });
+
+    context("when switching between projects", () => {
+      it("opens the selected project on the default branch", () => {
```
Comment (xla) [inline cypress/integration/project_source_browsing.spec.js:201]:
```suggestion
      it("opens the selected project on the default repository and branch", () => {
```

File: cypress/integration/project_source_browsing.spec.js [hunk 15]
```diff
@@ -6,7 +6,7 @@ before(() => {
 
 beforeEach(() => {
   cy.visit("./public/index.html#/profile/projects");
-  cy.contains("Monadic").click();
+  cy.contains("platinum").click();
```
Comment (xla) [inline cypress/integration/project_source_browsing.spec.js:9]: Did we change the fixtures?

File: proxy/src/coco/source.rs (hunk 10 above)
Comment (xla) [reply proxy/src/coco/source.rs:222]: @FintanH We want to scope by namespace (peer_id, "repository") whenever doing source queiries. Under what condition would we not need to be explicit about that scope when fetching a blob?

File: proxy/src/http/source.rs (hunk 9 above)
Comment (xla) [reply proxy/src/http/source.rs:300]: Too many peers in the house. May th real peer please stand up!

File: proxy/src/http/source.rs (hunk 11 above)
Comment (xla) [reply proxy/src/http/source.rs:893]: I think the heuristic is a bit more complicated and depends on the state of the local monorepo: https://www.notion.so/monadic/Overview-4969e31627984cc1938a6603d7d33738#3f74974beb934d81a911fc1eed1f8a88

File: ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte (hunk 12 above)
[the remark this answers is not in the document]
Comment (rudolfs) [reply ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte]: So I tried with the shareableIdentityIdentifier, but I feel the current design doesn't have enough space to give the user much meaningful information in the collapsed version:

Screenshot 2020-07-02 at 10 21 08

Slightly better in the expanded version:
Screenshot 2020-07-02 at 10 21 12

Isn't it enough that the avatar is different, hence it's derived from the PeerId?

File: proxy/src/http/source.rs (hunk 11 above)
Comment (rudolfs) [reply proxy/src/http/source.rs:893]: Here `peer` is `librad::net::peer::PeerApi`.

File: cypress/integration/project_source_browsing.spec.js (hunk 15 above)
Comment (rudolfs) [reply cypress/integration/project_source_browsing.spec.js:9]: I did. We created a project called "Monadic" before, which would make more sense for an org, so I changed it to something that resembles a project.

Comment (xla) [reply cypress/integration/project_source_browsing.spec.js:9]: Thanks for the clarification.

File: proxy/src/coco/source.rs (hunk 10 above)
Comment (FintanH) [reply proxy/src/coco/source.rs:222]: Discussed on IRC. I didn't provide enough context here. Basically what I was trying to say that I'm not sure whether `blob` et al. should be aware of `PeerId` or if we make a `Browser` remote aware (where the `PeerId` is the remote). I need to evaluate whether this is useful for surf (my feeling is that it is) and then see what changes we can make. This is can be left as a reminder to do that work ;)

File: proxy/src/http/source.rs (hunk 11 above)
Comment (FintanH) [reply proxy/src/http/source.rs:893]: Ya, but we're in tests so passing in `peer.peer_id`, knowing that this `peer` created the projects, seems like a reasonable test to me. But it was just a suggestion :)

File: ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte [hunk 16]
```diff
@@ -108,10 +131,11 @@
   hidden={expanded}>
   <div class="selector-avatar">
     <Avatar
-      title={revisions[0].identity.metadata.handle}
-      avatarFallback={revisions[0].identity.avatarFallback}
+      title={currentSelectedPeer.identity.metadata.handle}
```
Comment (FintanH) [inline ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte:132]: It's still `handle` :grimacing:

File: ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte (hunk 12 above)
[the remark this answers is not in the document]
Comment (xla) [reply ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte]: We should just shortne the peer_id. Is there a way to have a shortform similiar to how Oids have a 7-digiti version @FintanH?

[the remark this answers is not in the document]
Comment (FintanH) [reply ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte]: What if we have a shortened version of the shareable identifier that does this?

[the remark this answers is not in the document]
Comment (rudolfs) [reply ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte]: Ok, I'll add the full id for now, but I feel that this is a bit out of scope for this issue.
Would like to address the shortening in a follow-up once we have more of the real backend. Perhaps we'll also need to adjust the design of this dropdown anyway.

[the remark this answers is not in the document]
Comment (FintanH) [reply ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte]: Ok, can you make a follow-up ticket for this?

File: ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte (hunk 16 above)
Comment (rudolfs) [reply ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte]: https://github.com/radicle-dev/radicle-upstream/pull/620/commits/b7fe61d0bc3f0547a378b2c4d6b15657cb0ec679

File: ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte (hunk 12 above)
[the remark this answers is not in the document]
Comment (rudolfs) [reply ui/DesignSystem/Component/SourceBrowser/RevisionSelector.svelte]: https://github.com/radicle-dev/radicle-upstream/pull/620/commits/b7fe61d0bc3f0547a378b2c4d6b15657cb0ec679

File: proxy/src/http/source.rs (hunk 11 above)
Comment (rudolfs) [reply proxy/src/http/source.rs]: https://github.com/radicle-dev/radicle-upstream/pull/620/commits/105b9d048bd8d1b6bb8573f496246a6f00b048fb

**Comment by rudolfs** [comment]
@FintanH right, addressed here: https://github.com/radicle-dev/radicle-upstream/pull/620/commits/e819df8b04c9b1c33c8af7272c511b805ac427f1.
